pkg/roachpb/metadata.go

```diff
@@ -416,6 +416,12 @@ func (r ReplicaDescriptor) String() string {
 	return redact.StringWithoutMarkers(r)
 }
 
+// IsSame returns true if the two replica descriptors refer to the same replica,
+// ignoring the replica type.
+func (r ReplicaDescriptor) IsSame(o ReplicaDescriptor) bool {
+	return r.NodeID == o.NodeID && r.StoreID == o.StoreID && r.ReplicaID == o.ReplicaID
```

Comment:
What are the relations between node, store, and replica?

Reply:
Is `ReplicaID` local to a pair of (`NodeID`, `StoreID`)? Can it be compared without any of the two?

Reply:
One-to-many all the way down. One node can have many stores (disks), one store can have many replicas.

The node ID must be unique within the cluster. The store ID must be unique per node. The replica ID must be unique within the range. The range ID is sort of implied here, in that the replica descriptor is stored within a range descriptor.

In order to route a request to a replica we need to know all four. A replica ID won't move between nodes/stores -- if we have to move it, we'll create a new replica (with a new ID) on a different node, populate it, and then delete the old one.

Reply:
You're right in that we technically only need to compare the replica ID here, since we're operating within the context of a single range. We check the node ID and store ID just to make sure we have the right "address" for it, I suppose.

Reply:
Ok, so: (rangeID, replicaID) -> (nodeID, storeID), and rangeID is implied by the caller of `IsSame` (otherwise, due to locality of replicaID, it would be comparing apples to tables). Hence, is it enough to just compare `ReplicaID`s? From what you said it follows that if rangeID and replicaID are the same then nodeID and storeID are the same too.

Reply:
Ah, you answered this already. Were racing with the last 2 comments :)

Reply:
In principle it should be, yeah. But it's cheap enough to check the node and store too, so I don't see a strong reason not to. If there should be a mismatch for whatever reason, then there's no real point in trying to contact the replica anyway, because the transport will send it to the wrong place.